[feature] wrapping cost function to return best value explored

VolodyaCO opened the pull request:
## Description

Closes #29 

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

VolodyaCO commented:
It looks like mypy is failing due to this: https://github.com/python/mypy/issues/3060. Not sure what to do 🤷‍♂️

AthenaCaesura commented on src/orquestra/opt/optimizers/scipy_optimizer.py:
```diff
@@ -14,6 +14,77 @@
+            None,
+        ] = None,
+    ) -> None:
+        self.cost_function = cost_function
+        # Inherit all attributes from the cost function
+        for attr in dir(cost_function):
+            if not attr.startswith("__"):
+                setattr(self, attr, getattr(cost_function, attr))
```
When making a composition, it's usually best to get the attributes from the wrapped objects rather than just copying them over. This is because you can easily end up accidentally importing attributes that might change your execution. That is, your object won't be very consistent between instances. I would suggest that you really look to ensure there are no alternatives before doing this sort of thing.

VolodyaCO replied:
Yes, this can be a bit dangerous. I will perform a manual check for the time being so that there are no collisions with the methods and other attribute names.

AthenaCaesura commented on src/orquestra/opt/optimizers/scipy_optimizer.py:
```diff
@@ -14,6 +14,77 @@
+from .pso.continuous_pso_optimizer import _get_bounds_like_array
+
+
+class _CostFunctionWithBestValueType(type):
+    def __instancecheck__(cls, obj: object) -> bool:
+        if (
+            isinstance(obj, Callable)
+            and hasattr(obj, "best_value")
+            and hasattr(obj, "best_params")
+        ):
+            return True
+        return False
```
This is a nice little piece of code! But I'm afraid it might be needlessly opaque since isinstance is only used once here and just for an assert statement. metaclasses are considered "heavy machinery" so it's best not to use them when they could be replaced with a simple if statement.

VolodyaCO replied:
If I remove this to check the new instance check, I get some linting problems 😢

AthenaCaesura replied:
Yes, but since you only use this isinstance once, it's not really worth changing the __instancecheck__ method right?

VolodyaCO replied:
I had to do this trick for the tensor train optimiser PR too #31. Do you want me to instead have a custom `isinstanceCostFunctionWithBestValue`, or something like that?

AthenaCaesura replied:
No, it's probably fine. Not worth holding up the PR for this.

AthenaCaesura commented:
Can we add an option for the user to go back to the old way of using the scipy optimizer? People might get confused if we dont' 😅

AthenaCaesura commented on src/orquestra/opt/optimizers/scipy_optimizer.py:
```diff
@@ -66,7 +142,7 @@ def _minimize(
 
         """
-
+        assert isinstance(cost_function, _CostFunctionWithBestValue)
```
Can't we just replace this with type checking in the cost_function input? Not sure why it needs a check here. At most, I would just put an easy to understand error wherever the first conflict arises that requires one to use a `CostFunctionWithBestValue`.

VolodyaCO replied:
I removed this and performed an instance type check instead to be able to deduce which attributes I should access to get the best parameters and function value.

AthenaCaesura commented on src/orquestra/opt/optimizers/scipy_optimizer.py:
```diff
@@ -14,6 +14,77 @@
+    def __init__(
+        self,
+        cost_function: Union[CallableWithGradient, Callable],
+        constraints: Sequence[Dict[str, Callable]],
```
```suggestion
        constraints: Tuple[Dict[str, Callable]],
```
Just to keep typing consistent between optimizers. 😉

AthenaCaesura replied:
If it is possible to have a `CostFunctionWithBestValue` with no constraints, it might be best to put the `Optional` type on there too.

VolodyaCO replied:
Done 😄

AthenaCaesura commented on tests/orquestra/opt/optimizers/scipy_optimizer_test.py:
```diff
@@ -107,3 +108,30 @@ def test_SLSQP_with_inequality_constraints(self):
         )
         assert results_with_constraints.opt_params.sum() >= 3
+
+    # https://github.com/scipy/scipy/issues/17673 reports a function and that is
+    # minimised with an initial point that has a given cost function value which is
+    # lower than the optimised value by the Powell optimizer. This test is to ensure
+    # that our solution over scipy is not affected by this issue.
```
Nice! Good testing!

VolodyaCO commented:
I'm using this in a GEO workflow but it seems as though it isn't working as expected. Here are the logs from the workflow:
```zsh
{"timestamp": "2023-01-25 08:32:15,166", "level": "INFO", "filename": "history_selector.py:102", "message": Number of pre-selected candidates 398. The radius is 0.0390625}
{"timestamp": "2023-01-25 08:32:15,298", "level": "INFO", "filename": "base.py:769", "message": Probs sum is: 1.0000000000000022}
{"timestamp": "2023-01-25 08:32:15,298", "level": "INFO", "filename": "base.py:770", "message": xtrain size is: 7924}
{"timestamp": "2023-01-25 08:32:15,298", "level": "INFO", "filename": "base.py:771", "message": Number of observations are: 7924}
{"timestamp": "2023-01-25 08:32:15,298", "level": "INFO", "filename": "base.py:772", "message": Minimum cost is: -8.103484550314727}
Composition labels in HDBSCAN clustering:
Counter({-1: 4496, 4: 1412, 8: 838, 7: 283, 1: 197, 5: 194, 6: 161, 0: 121, 3: 113, 2: 109})
Cleared cache for LangerInCorners(m=4) in 2 dimensions and seed 0
{"timestamp": "2023-01-25 08:32:16,405", "level": "INFO", "filename": "geo_workflows.py:315", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Finished in 0.024 seconds"}}
{"timestamp": "2023-01-25 08:32:16,405", "level": "INFO", "filename": "geo_workflows.py:316", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Optimum value: -7.866315123621101"}}
{"timestamp": "2023-01-25 08:32:16,433", "level": "INFO", "filename": "local_workflows.py:62", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :::: Cleaned history from length 30 to length 1. dl = 0.1810866855724538"}}
Cleared cache for LangerInCorners(m=4) in 2 dimensions and seed 0
{"timestamp": "2023-01-25 08:32:17,202", "level": "INFO", "filename": "geo_workflows.py:315", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Finished in 0.025 seconds"}}
{"timestamp": "2023-01-25 08:32:17,203", "level": "INFO", "filename": "geo_workflows.py:316", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Optimum value: -7.8663151681677315"}}
{"timestamp": "2023-01-25 08:32:17,227", "level": "INFO", "filename": "local_workflows.py:62", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :::: Cleaned history from length 67 to length 2. dl = 0.17779190436107137"}}
Cleared cache for LangerInCorners(m=4) in 2 dimensions and seed 0
{"timestamp": "2023-01-25 08:32:17,966", "level": "INFO", "filename": "geo_workflows.py:315", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Finished in 0.024 seconds"}}
{"timestamp": "2023-01-25 08:32:17,966", "level": "INFO", "filename": "geo_workflows.py:316", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Optimum value: -7.866315173534457"}}
{"timestamp": "2023-01-25 08:32:17,992", "level": "INFO", "filename": "local_workflows.py:62", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :::: Cleaned history from length 67 to length 1. dl = 0.17560887003556275"}}
Cleared cache for LangerInCorners(m=4) in 2 dimensions and seed 0
{"timestamp": "2023-01-25 08:32:18,650", "level": "INFO", "filename": "geo_workflows.py:315", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Finished in 0.024 seconds"}}
{"timestamp": "2023-01-25 08:32:18,651", "level": "INFO", "filename": "geo_workflows.py:316", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Optimum value: -7.856305814156559"}}
{"timestamp": "2023-01-25 08:32:18,680", "level": "INFO", "filename": "local_workflows.py:62", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :::: Cleaned history from length 62 to length 2. dl = 0.17537807662467564"}}
Cleared cache for LangerInCorners(m=4) in 2 dimensions and seed 0
{"timestamp": "2023-01-25 08:32:19,406", "level": "INFO", "filename": "geo_workflows.py:315", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Finished in 0.024 seconds"}}
{"timestamp": "2023-01-25 08:32:19,407", "level": "INFO", "filename": "geo_workflows.py:316", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :: Optimum value: -7.856305810418254"}}
{"timestamp": "2023-01-25 08:32:19,431", "level": "INFO", "filename": "local_workflows.py:62", "message": {"run_id": "wf.local_run.0000000@tsk.task_local-0.00000", "logs": "     :::: Cleaned history from length 60 to length 2. dl = 0.17537775444008236"}}
```
From GEO optimisation we get a minimum of `-8.something`. The point corresponding to this value in the cost function is set as the initial parameter of one of the local optimisation runs, which return as best value `-7.8something`. This is not the expected behaviour.

AthenaCaesura commented:
Let me know if you need any help with debugging and GEO to get the behavior you're expecting!
